### ISSUE TITLE
OpenStack provider: high load on Nova Compute API

### ISSUE THREAD

**Issue by MatMaul** [kubernetes/kubernetes#49867]
When using a lot of Kubernetes nodes with the OpenStack provider the load on the Nova Compute API (and possibly depending one like Neutron) is high.
This is because each kubelet makes a request to the API every 10s to fetch the IPs of the VM.

After having a look at the other providers it seems like they generally try to use the local metadatas to find this info, no API call seems to be made.
I tried to do that with OpenStack, however even with recent versions of OpenStack metadatas the details of the networks are not available : I can have the IP and the network ID but I can't tell directly if this is a public (shared) network.
On top of that a lot of Nova deployment like ours use a metadata server instead of a drive and the load problem would be the same in this case.

I fall back to just cache the result of the API call in the kubelet. It works but it would not take into account a change of floating IP for example. Perhaps this cache should be invalidate every hour (at a random time to distribute the load).

I'll push the patch in a PR, feedback welcome.

**Comment by MatMaul** [comment]
@dims @FengyunPan if you have any comments on that.

Regarding the fact that we would not see a change of floating IP with the cache I think it would be the same for most existing providers since they use local metadatas, so perhaps it is ok to do so.

**Comment by MatMaul** [comment]
up @anguslees @dims @FengyunPan 
Any feedback on that ? Shoulfd I send my PR for caching the metadatas ?

**Comment by dims** [comment]
@MatMaul yes, please send the PR for caching the metadata, will help with the discussion i think

**Comment by anguslees** [comment]
Cache invalidation becomes an issue with any cache of mutable data.  Also, from the k8s side, all _we_ can do is change the point at which the sum of all the metadata queries becomes an issue - we can't change the _shape_ of the scalability curve..   So we're not going to find any good answer here.

I'm inclined to reassign this bug to openstack as the right place to fix it.  It seems to me like a _design requirement_ that the openstack metadata server capacity should increase in line with the number of hosts.  (Perhaps by caching metadata on the underlying openstack hosts?  At least inside openstack they would be able to accurately work out when to invalidate that cache)

Thoughts?

**Comment by rochaporto** [comment]
Scaling on openstack is already possible in two ways: using cells (spread the load) and using a memcache layer in front of the nova metadata api.

**Comment by MatMaul** [comment]
I understand your point, we can indeed (and will) increase metadata_cache_expiration in Nova (for ref if someone has the same issue).

However I still think that decoupling the update of the ips from the kubelet heartbeat could be helpful, doing so every 10s is quite overkill. With a randomized start so that all the requests don't end up at the same time.
It can be either a change in k8s core and a new global option, or a cache in the openstack provider itself.

We will probably be able to live without it once the cache on the metadata server is increased, so let me know if you think this would be useful/accepted. I am volunteering to do the code of course :)

**Comment by MatMaul** [comment]
If not this issue can be closed BTW.

**Comment by anguslees** [comment]
Thanks @MatMaul.  I agree that separating the heartbeat code from "what are your IPs" makes sense.  I can't come up with a good effective alternative for the latter however - if you extend the polling period, then you just extend the limbo period following a change to the node's IP address (and eg this is/was a problem in neutron with the dhcp lease renewal time, for the same reasons).

Afaics we need to either give up and not support changing the IP without a reboot (perhaps this would be ok?), or have an event-driven "push" flow.  Anything in the middle ends up with a result that is the worst of both worlds: still does repeated queries _and_ changes don't propagate for some inconveniently long time. The event-driven flow doesn't fit well with the existing controller<->node REST model, so instead we poll frequently and rely on some sort of "cache with event-driven invalidation" (ie: an openstack-side metadata cache) in order to make that not terrible and look like the event driven flow.

Having said all that, an alternative that might make sense in the openstack provider is to say that it doesn't _matter_ what the openstack API thinks the addresses are - what matters are the addresses configured on the host.  So we _could_ just query the local interfaces from the kubelet without hitting the metadata api at all (for the kubelet use case, istr there's another codepath that gets called from the controller too?).  This basically moves the problem to the openstack DHCP cache/expiry period, which again doesn't make it go away but makes it more obvious that we should blame someone else at this point :stuck_out_tongue:  The additional complications with this version are knowing which interfaces/addresses are "real" and which are (eg) docker virtual interfaces, etc - but we can put enough heuristics in place to make that work most of the time.

TL;DR: I think there are changes you could make that would be accepted, and would address your immediate issue.  Personally, I think they're dealing with the symptom rather than cause, and you're better off spending your time adjusting metadata_cache_expiration in nova :stuck_out_tongue:

**Comment by MatMaul** [comment]
@anguslees actually I don't really understand the use case where the internal IP would change. The external one I get it it is floating IP that can change anytime, however I guess the most important one is the internal one right ? You are talking Neutron DHCP lease, I don't think Neutron would change a private IP on the fly ?
I think we are fine keeping it like that for now in k8s, but I am trying to understand what issues you saw regarding IP changing in OpenStack and its consequence on a k8s cluster, so we can adjust the metadata cache value accordingly ;)

**Comment by dims** [comment]
@MatMaul looking forward to your "PR for caching the metadatas" :)

**Comment by anguslees** [comment]
On Tue, 19 Sep 2017, 22:18 Mathieu Velten <notifications@github.com> wrote:

> @anguslees <https://github.com/anguslees> actually I don't really
> understand the use case where the internal IP would change. The external
> one I get it it is floating IP that can change anytime, however I guess the
> most important one is the internal one right ? You are talking Neutron DHCP
> lease, I don't think Neutron would change a private IP on the fly ?
>
I'm thinking of things like changing the neutron port IP and the effect
that downstream caching has on that transition.  See this thread  for how
this problem affected neutron itself:
http://lists.openstack.org/pipermail/openstack-dev/2015-January/055360.html

>
>

**Comment by rochaporto** [comment]
@anguslees would the 'use the internal IP instead of metadata reply' be a better option? I'm willing to implement this.

**Comment by MatMaul** [comment]
@dims I didn't send it since it looks that we agreed it was not going to be integrated no ? The patch was very simple, I didn't even put an expiration to the cache at the time:
https://github.com/MatMaul/kubernetes/commit/a659d027a5a165c6c478e446f0d07c0da3438d7f

@rochaporto if we get a ok for that I can do it or help, let me know.
